### PR TITLE
Implement NonEmptyList.:: and prepended

### PR DESCRIPTION
**Pull request by adamgfraser**

### CODE REVIEW

**Review by sideeffffect — COMMENTED** [review]
Looks like a nice addition :+1: 
But note that it needs formatting and some fix for 2.11 (doesn't compile) and 2.12 (the compiler crashes :scream_cat: )

**Comment by adamgfraser** [comment]
`prepended` doesn't exist on Scala 2.11 or 2.12.

**Review by sideeffffect — APPROVED** [review]
Nice, let's get this merged :rocket:

File: core/shared/src/main/scala/zio/prelude/NonEmptyList.scala [hunk 1]
```diff
@@ -407,8 +413,11 @@ sealed trait NonEmptyList[+A] { self =>
   /**
    * Converts this `NonEmptyList` to the `::` case of a `List`.
    */
-  final def toCons[A1 >: A]: ::[A1] =
-    reduceMapRight[::[A1]](::(_, Nil))(::(_, _))
+  final def toCons[A1 >: A]: ::[A1] = {
```
[the remark this answers is not in the document]
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/NonEmptyList.scala:416]: It is already implicitly convertible to that and I think the cons and single cases are how you want to interact with it if you are interacting with it if you are interacting with it as a nonempty list.

[the remark this answers is not in the document]
Comment (sideeffffect) [reply core/shared/src/main/scala/zio/prelude/NonEmptyList.scala:416]: You mean for pattern matching? But those things aren't mutually exclusive, right?
And converting to regular `List` would be just _O(n)_.

[the remark this answers is not in the document]
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/NonEmptyList.scala:416]: Yeah I'll take a look at that.

[the remark this answers is not in the document]
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/prelude/NonEmptyList.scala:416]: You lose exhaustivity checking on the pattern matching, which I think is quite a loss.

[the remark this answers is not in the document]
Comment (sideeffffect) [reply core/shared/src/main/scala/zio/prelude/NonEmptyList.scala:416]: Aaaah, yes. That's a good point. We could simulate this by returning `Either` and matching on that, but that's just not as elegant...

**Pull request merged**